atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/transaction/impl/expectations/TrackingKeyValueServiceForwardingTest.java

```diff
@@ -0,0 +1,326 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+/**
+ * Broadly tests that {@link TrackingKeyValueServiceImpl} methods forward delegate results using: physical equality
+ * when relevant, and element/component physical equality when the semantics are different (methods returning
+ * iterators or collections of iterators).
+ * Mocks are used whenever possible. Exceptions include: types which cannot be mocked (e.g. byte arrays),
+ * interfaces/implementations annotated with {@link org.mockito.DoNotMock} or similar (e.g. {@link Multimap}), and
+ * when the method semantics and resulting tests impose it (e.g. methods where an iterator is wrapped can only be tested
+ * by equality/in-order testing of its components, {@link TrackingKeyValueServiceImpl#getCandidateCellsForSweeping}).
+ */
+@RunWith(MockitoJUnitRunner.class)
+public final class TrackingKeyValueServiceForwardingTest {
+    private static final long TIMESTAMP = 12L;
+    private static final byte[] BYTES_1 = new byte[1];
+    private static final byte[] BYTES_2 = new byte[2];
+    private static final Cell CELL_1 = Cell.create(BYTES_1, BYTES_1);
+    private static final Cell CELL_2 = Cell.create(BYTES_2, BYTES_2);
+    private static final Value VALUE_1 = Value.create(BYTES_1, TIMESTAMP);
+    private static final Value VALUE_2 = Value.create(BYTES_2, TIMESTAMP);
+
+    @Mock
+    private TableReference tableReference;
+
+    @Mock
+    private RangeRequest rangeRequest;
+
+    @Mock
+    private List<byte[]> rows;
+
+    @Mock
+    private Map<Cell, Long> timestampByCellMap;
+
+    @Mock
+    private Map<Cell, Value> valueByCellMap;
+
+    @Mock
+    private KeyValueService delegate;
+
+    private TrackingKeyValueService trackingKvs;
+
+    @Before
+    public void setUp() {
+        trackingKvs = new TrackingKeyValueServiceImpl(delegate);
+    }
+
+    @Test
+    public void getAsyncWrapsAndForwardsDelegateFutureResult() throws ExecutionException, InterruptedException {
+        when(delegate.getAsync(tableReference, timestampByCellMap)).thenReturn(Futures.immediateFuture(valueByCellMap));
+
+        assertThat(trackingKvs.getAsync(tableReference, timestampByCellMap).get())
+                .isSameAs(valueByCellMap);
+    }
+
+    @Test
+    public void getAsyncResultCompletesIffDelegateResultCompletes() {
+        SettableFuture<Map<Cell, Value>> delegateFuture = SettableFuture.create();
+        when(delegate.getAsync(tableReference, timestampByCellMap)).thenReturn(delegateFuture);
+
+        ListenableFuture<Map<Cell, Value>> future = trackingKvs.getAsync(tableReference, timestampByCellMap);
+        assertThat(future).isNotDone();
+
+        // completes the delegate future
+        delegateFuture.set(ImmutableMap.of());
```

Comment:
nit: may as well set the value to something we can then check afterwards as well!